dali/pipeline/data/tensor_vector.cc

```diff
@@ -60,20 +65,170 @@ TensorVector<Backend>::TensorVector(TensorVector<Backend> &&other) noexcept {
+  DALI_ENFORCE(GetLayout() == layout,
+               make_string("Sample must have the same layout as a target batch current: ",
+                           GetLayout(), " new: ", layout, " for ", sample_idx, "."));
+  DALI_ENFORCE(is_pinned() == pinned,
+               make_string("Sample must have the same pinned status as target batch, current: ",
+                           is_pinned(), " new: ", pinned, " for ", sample_idx, "."));
+  SetContiguous(false);
+  // Setting a new share overwrites the previous one - so we can safely assume that even if
+  // we had a sample sharing into TL, it will be overwritten
+  tensors_[sample_idx]->ShareData(ptr, bytes, pinned, shape, type, order);
+  tl_->Reset();
+}
+
+template <typename Backend>
+void TensorVector<Backend>::UnsafeCopySample(int sample_idx, const TensorVector<Backend> &src,
+                                             int src_sample_idx, AccessOrder order) {
+  // TODO(klecki): more consistency checks, contiguous -> non-contiguous removes shares_data from
+  // samples
+  // Bounds check
+  assert(sample_idx >= 0 && sample_idx < static_cast<int>(curr_tensors_size_));
+  assert(src_sample_idx >= 0 && src_sample_idx < static_cast<int>(src.curr_tensors_size_));
+  DALI_ENFORCE(type() == src.type(),
+               make_string("Sample must have the same type as a target batch, current: ", type(),
+                           " new: ", src.type(), " for ", sample_idx, " <- ", src_sample_idx, "."));
+  DALI_ENFORCE(sample_dim() == src.shape().sample_dim(),
+               make_string("Sample must have the same dimensionality as a target batch, current: ",
+                           sample_dim(), " new: ", src.shape().sample_dim(), " for ", sample_idx,
+                           " <- ", src_sample_idx, "."));
+  DALI_ENFORCE(
+      GetLayout() == src.GetLayout(),
+      make_string("Sample must have the same layout as a target batch current: ", GetLayout(),
+                  " new: ", src.GetLayout(), " for ", sample_idx, " <- ", src_sample_idx, "."));
+
+  // Either the shape matches and we can copy data as is or the target is just an individual sample
+  bool can_copy = tensors_[sample_idx]->shape() == src.tensors_[src_sample_idx]->shape() ||
+                  (!tl_->has_data() && state_ == State::noncontiguous);
+
+  DALI_ENFORCE(
+      can_copy,
+      "Copying samples into TensorVector can happen either for exact shape match or when the "
+      "TensorVector is truly non contiguous. Either Resize first to the desired shape or reset the "
+      "TensorVector and SetSize for desired number of samples in non-contiguous mode.");
+
+  tensors_[sample_idx]->Copy(*src.tensors_[src_sample_idx], order);
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetupLike(const Tensor<Backend> &sample) {
+  DALI_ENFORCE(!has_data(),
+               "Batch object can be initialized this way only when it doesn't have allocation.");
```

Comment:
```suggestion
               "Batch object can be initialized this way only when it doesn't have an allocation.");
```
or (probably better):
```suggestion
               "Batch object can be initialized this way only when it isn't allocated");
```

Reply:
done